src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,136 @@
+            {
+                foreach (string nugetNamespace in configuration.Namespaces[packageSourceKey])
+                {
+                    Add(packageSourceKey, nugetNamespace);
+                }
+            }
+        }
+
+        private void Add(string packageSourceKey, string namespaceId)
+        {
+            SearchNode currentNode = _root;
+
+            if (namespaceId == null)
+            {
+                throw new ArgumentNullException(nameof(namespaceId));
+            }
+
+            // To prevent from unwanted behaviour.
+            if (namespaceId.Length > PackageNamespacesConfiguration.PackageIdMaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namespaceId));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            packageSourceKey = packageSourceKey.ToLowerInvariant().Trim();
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    currentNode.IsGlobbing = true;
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    currentNode.Children[c] = new SearchNode();
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (string.IsNullOrEmpty(currentNode.NamespaceId))
+            {
+                currentNode.NamespaceId = namespaceId;
+            }
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+            }
+
+            currentNode.PackageSources.Add(packageSourceKey);
+        }
+
+        /// <summary>
+        /// Get package source names with matching prefix "term" from package namespaces section.
+        /// </summary>
+        /// <param name="term">Search term. Never null. </param>
+        /// <returns>Package source names with matching prefix "term" from package namespaces.</returns>
+        /// <exception cref="ArgumentNullException"> if <paramref name="term"/> is null or empty.</exception>
+        public HashSet<string> GetConfiguredPackageSources(string term)
+        {
+            if (string.IsNullOrWhiteSpace(term))
+            {
+                throw new ArgumentNullException(nameof(term));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            term = term.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+            SearchNode currentNode = _root;
+            SearchNode longestMatchingPrefixNode = null;
```

Comment:
I think `longestMatchingPrefixNode` should be a collection of nodes. Let's says we have following namespaces defined.

1. `NuGet.*` packages should be fetched from `source1`
2. NuGet.Common*` packages should be fetched from `source2`.

While searching for `NuGet.Common1` package, I think the Trie structure should return both `source1` and `source2` because the package Id satisfies both namespaces. The current logic returns only `source2`.

Reply:
We match to longest matching prefix.
Please check case 1C from [spec](https://github.com/NuGet/Home/blob/dev/proposed/2021/PackageNamespaces.md).
![image](https://user-images.githubusercontent.com/8766776/122071408-90f4dc00-cdab-11eb-8eb2-e68acef17942.png)